File: part-02/src/test/java/eu/luminis/workshop/smallsteps/logic/domainservice/LegoStockReturningTest.java
package eu.luminis.workshop.smallsteps.logic.domainservice;

import eu.luminis.workshop.smallsteps.logic.domainservice.auth.AccessDeniedException;
import eu.luminis.workshop.smallsteps.logic.domainservice.helper.SetupLegoTestApp;
import eu.luminis.workshop.smallsteps.logic.domainservice.state.LegoBox;
import eu.luminis.workshop.smallsteps.logic.domainservice.state.LegoSetForBuilder;
import eu.luminis.workshop.smallsteps.logic.domainservice.state.StockState;
import eu.luminis.workshop.smallsteps.persistence.InMemoryLegoPartCatalog;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.Collections;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;

public class LegoStockReturningTest {
    SetupLegoTestApp app;

    @BeforeEach
    void setUp() {
        app = new SetupLegoTestApp();
    }

    @Test
    public void should_register_returning_complete_lego_set() {
        StockState initialState = new StockState(
                null,
                null,
                List.of(new LegoSetForBuilder(app.millenniumFalcon, app.harry)),
                null,
                null
        );
        LegoStock handler = new LegoStock(app.harryAuth, new InMemoryLegoPartCatalog(), initialState);

        StockState newState = handler.returnComplete(app.millenniumFalcon);

        assertThat(newState.getAtBuilder()).isEqualTo(newState.getAtBuilder());
    }

    @Test
    public void should_fail_to_register_returning_lego_set_when_not_at_builder() {
        StockState initialState = new StockState(
                null,
                null,
                Collections.emptyList(),
                null,
                null
        );
        LegoStock handler = new LegoStock(app.harryAuth, new InMemoryLegoPartCatalog(), initialState);

        assertThatThrownBy(() -> {
            handler.returnComplete(app.millenniumFalcon);
        }).isInstanceOf(IllegalStateException.class)
                .hasMessage("Cannot return " + app.millenniumFalcon.getNumber() + " as builder " +
                        app.harry.toString() + " does not have it.");

        assertThatThrownBy(() -> {
            handler.returnInComplete(new LegoBox(app.millenniumFalcon, 42, app.missingParts));
        }).isInstanceOf(IllegalStateException.class)
                .hasMessage("Cannot return " + app.millenniumFalcon.getNumber() + " as builder " +
                        app.harry.toString() + " does not have it.");

    }

    @Test
    public void should_fail_to_register_returning_lego_set_when_not_at_builder_but_is_at_different_builder() {
        StockState initialState = new StockState(
                null,
                null,
                List.of(new LegoSetForBuilder(app.millenniumFalcon, app.sally)),
                null,
                null
        );
        LegoStock handler = new LegoStock(app.harryAuth, new InMemoryLegoPartCatalog(), initialState);

        assertThatThrownBy(() -> {
            handler.returnComplete(app.millenniumFalcon);
        }).isInstanceOf(IllegalStateException.class)
                .hasMessage("Cannot return " + app.millenniumFalcon.getNumber() + " as builder " +
                        app.harry.toString() + " does not have it.");

        assertThatThrownBy(() -> {
            handler.returnInComplete(new LegoBox(app.millenniumFalcon, 42, app.missingParts));
        }).isInstanceOf(IllegalStateException.class)
                .hasMessage("Cannot return " + app.millenniumFalcon.getNumber() + " as builder " +
                        app.harry.toString() + " does not have it.");
    }

    @Test
    public void should_fail_to_register_returning_lego_set_when_not_autehnticated_as_builder() {
        StockState initialState = new StockState(
                null,
                null,
                List.of(new LegoSetForBuilder(app.millenniumFalcon, app.harry)),
                null,
                null
        );
        LegoStock handler = new LegoStock(app.bussumAuth, new InMemoryLegoPartCatalog(), initialState);

        assertThatThrownBy(() -> {
            handler.returnComplete(app.millenniumFalcon);
        }).isInstanceOf(AccessDeniedException.class);

        assertThatThrownBy(() -> {
            handler.returnInComplete(new LegoBox(app.millenniumFalcon, 42, app.missingParts));
        }).isInstanceOf(AccessDeniedException.class);
    }
}

File: part-02/src/test/kotlin/eu/luminis/workshop/smallsteps/logic/domainService/LegoStock - ReturningTest.kt
package eu.luminis.workshop.smallsteps.logic.domainService

import eu.luminis.workshop.smallsteps.logic.domainService.helper.LegoStockHelper.millenniumFalcon
import eu.luminis.workshop.smallsteps.logic.domainService.auth.AccessDeniedException
import eu.luminis.workshop.smallsteps.logic.domainService.helper.LegoStockHelper
import eu.luminis.workshop.smallsteps.logic.domainService.helper.TestUsers
import eu.luminis.workshop.smallsteps.logic.domainService.state.LegoBox
import eu.luminis.workshop.smallsteps.logic.domainService.state.StockState
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

@Suppress("ClassName")
internal class `LegoStock - ReturningTest` {
    private val missingParts = mapOf(
        "3898b" to 1,
        "818622" to 3,
    )

    @Test
    fun `should register returning complete lego set`() {
        // given
        val initialState = StockState(
            atBuilder = listOf(
                millenniumFalcon to TestUsers.harry
            ),
        )
        val handler = LegoStockHelper.build(initialState)

        // when
        val newState = handler.returningComplete(millenniumFalcon)

        // then
        assertEquals(initialState, newState)
    }

    @Test
    fun `should fail to register returning lego set when not at builder`() {
        // given
        val initialState = StockState(
            atBuilder = emptyList(),
        )
        val handler = LegoStockHelper.build(initialState)

        // then
        assertFailsWith<IllegalStateException> {
            // when
            handler.returningComplete(millenniumFalcon)
        }

        // and
        assertFailsWith<IllegalStateException> {
            // when
            handler.returningInComplete(LegoBox(millenniumFalcon, 42, missingParts))
        }
    }

    @Test
    fun `should fail to register returning lego set when not at builder but is at different builder`() {
        // given
        val initialState = StockState(
            atBuilder = listOf(
                millenniumFalcon to TestUsers.sally
            ),
        )
        val handler = LegoStockHelper.build(initialState)

        // then
        assertFailsWith<IllegalStateException> {
            // when
            handler.returningComplete(millenniumFalcon)
        }

        // and
        assertFailsWith<IllegalStateException> {
            // when
            handler.returningInComplete(LegoBox(millenniumFalcon, 42, missingParts))
        }
    }

    @Test
    fun `should fail to register returning lego set when not authenticated as builder`() {
        // given
        val initialState = StockState(
            atBuilder = listOf(
                millenniumFalcon to TestUsers.harry
            ),
        )
        val handler = LegoStockHelper.build(initialState, auth = TestUsers.bussumAuth)

        // then
        assertFailsWith<AccessDeniedException> {
            // when
            handler.returningComplete(millenniumFalcon)
        }

        // and
        assertFailsWith<AccessDeniedException> {
            // when
            handler.returningInComplete(LegoBox(millenniumFalcon, 42, missingParts))
        }
    }
}